subsys/pm/device_runtime.c

```diff
@@ -601,6 +601,109 @@ int pm_device_runtime_usage(const struct device *dev)
 	return dev->pm_base->usage;
 }
 
+void pm_device_runtime_reference_init(struct pm_device_runtime_reference *ref)
+{
+	k_sem_init(&ref->lock, 1, 1);
+	ref->active = false;
+}
+
+static bool reference_trylock(struct pm_device_runtime_reference *ref)
+{
+	if (k_is_in_isr() && k_sem_take(&ref->lock, K_NO_WAIT)) {
+		return false;
+	}
+
+	(void)k_sem_take(&ref->lock, K_FOREVER);
+	return true;
+}
+
+static void reference_unlock(struct pm_device_runtime_reference *ref)
+{
+	k_sem_give(&ref->lock);
+}
```

Comment:
It seems like a semaphore might be overkill here, if it's just guarding access to the bool as a critical section, could it just be a spinlock instead? Or are you thinking that the `pm_device_runtime_get` will take a long time due to traversing domain callbacks?